plat/common/arm/lcpu.c

```diff
@@ -0,0 +1,250 @@
+	 *  starting secondary CPUs. We thus do the translation here once and
+	 *  cache the result.
+	 */
+	lcpu_start_paddr = ukplat_virt_to_phys(lcpu_start);
+
+	bsp_cpu_id = lcpu_arch_id();
+	uk_pr_info("Bootstrapping processor has the ID %ld\n",
+		   bsp_cpu_id);
+
+	/* Search for CPUS node in DTB */
+	fdt_cpu = fdt_path_offset(dtb, "/cpus");
+	if (unlikely(fdt_cpu < 0)) {
+		uk_pr_err("cpus node is not found in device tree\n");
+		return -EINVAL;
+	}
+
+	/* Get address cells */
+	naddr_prop = fdt_getprop(dtb, fdt_cpu, "#address-cells", &cell);
+	if (naddr_prop == NULL) {
+		naddr = FDT_ADDR_CELLS_DEFAULT;
+		uk_pr_warn("Using default value 2 for cpu address-cells.\n");
+	} else {
+		naddr = fdt32_to_cpu(naddr_prop[0]);
+	}
+	if (unlikely(naddr < 0 || naddr >= FDT_MAX_NCELLS)) {
+		uk_pr_err("Invalid address-cells size.\n");
+		return -EINVAL;
+	}
+	if (unlikely(naddr > FDT_ADDR_CELLS_DEFAULT)) {
+		uk_pr_err("Address-cells size greater than 2 not supported yet.\n");
+		return -EINVAL;
+	}
+
+	/* Get size cells */
+	nsize_prop = fdt_getprop(dtb, fdt_cpu, "#size-cells", &cell);
+	if (nsize_prop == NULL) {
+		nsize = FDT_SIZE_CELLS_DEFAULT;
+		uk_pr_warn("Using default value of 0 for cpu size-cells.\n");
+	} else {
+		nsize = fdt32_to_cpu(nsize_prop[0]);
+	}
+	/**
+	 *  According to the device-tree v0.3 specification,
+	 *  for the `size-cells` property of the `/cpus` node:
+	 *  "Value shall be 0. Specifies that no size is
+	 *  required in the reg property in children of this node."
+	 */
+	if (unlikely(nsize != 0)) {
+		uk_pr_err("Invalid size-cells value for cpu node.\n");
+		return -EINVAL;
+	}
+
+	/* Search all the CPU nodes in DTB */
+	fdt_for_each_subnode(subnode, dtb, fdt_cpu) {
+		int prop_len = 0;
+
+		cpu_prop = fdt_get_property(dtb, subnode,
+					    "device_type", &prop_len);
+		if (!cpu_prop)
+			continue;
+		if (strcmp(cpu_prop->data, "cpu"))
+			continue;
+
+		cpu_prop = fdt_get_property(dtb, subnode, "reg", &prop_len);
+		if (unlikely(cpu_prop == NULL || prop_len <= 0)) {
+			uk_pr_err("CPU node does not contain reg field "
+				  "(cpu_id), this core will not be enabled.\n");
+			continue;
+		}
+		id_reg = (const fdt32_t *) cpu_prop->data;
+		cpu_id = fdt_reg_read_number(id_reg, naddr);
+
+		cpu_prop = fdt_get_property(dtb, subnode,
+					    "enable-method", NULL);
+		if (!cpu_prop) {
+			if (cpu_id != bsp_cpu_id) {
+				uk_pr_err("enable-method for core 0x%lx not "
+					  "set, core will not be enabled.",
+					  cpu_id);
+				continue;
+			} else {
+				uk_pr_warn("enable-method for core 0x%lx not "
+					   "set\n", cpu_id);
+				bsp_found = 1;
+				continue;
+			}
+		} else if (strcmp(cpu_prop->data, "psci")) {
+			uk_pr_err("enable-method for core 0x%lx is not PSCI : "
+				  "(%s), core will not be enabled.\n", cpu_id,
+				  cpu_prop->data);
+			continue;
+		}
+
+		if (cpu_id != bsp_cpu_id) {
+			lcpu = lcpu_alloc(cpu_id);
+			if (unlikely(!lcpu)) {
+				uk_pr_warn("Maximum number of cores exceeded.\n");
+				break;
```

Comment:
I think this is a bug. It somewhat makes sense to check if we found the BSP for sanity purposes if _ALL_ cores are added. But if we abort early due to reaching the maximum number of cores, the assertation might fail. For example, it would be ok to have core 8 as BSP and discover only cores 0, 1, and 2 if the unikernel is configured for 4 cores.

Reply:
@marcrittinghaus You are right, I think we can just return if the maximum number of cores has been reached here.